feat: add command headings feature

jerusdp opened the pull request:
Implementing issue #1553

### Scope of this PR

The objective for this PR is to provide a facility for the user to place their commands under a custom header to allow the grouping of commands such as might be seen in `git --help`. 

The PR will not implement `CommandAction` which would be required to permit the user to replace the `help` command with a custom user `help` command which could be place under a custom header. The user can place `help` under a custom header by amending the default title using `Command::subcommand_help_heading` placing the help command and the updated section at the top of the list of commands. 

The will not change the scope of the  `Command::next_help_heading` to include processing for commands as this is a breaking change. 

## Design Statements

| Statement | Comment |
| --- | --- |
| MUST provide API to specify a custom section header for a user command | The key objective of the change |
| MUST keep commands with the same header together | |
| MUST display commands with no custom header under the default header | |
| MUST display help command under the standard header | To amend this the user should change the default header |
| WILL NOT change scope of `Command::next_help_heading` to include subcommands | | 
| WILL NOT deprecate `Command::subcommand_help_heading` | It can be used to generate custom section heading for help |
| WILL NOT implement custom replacement of help | Future solution requiring `CommandAction` 
<!--
Thanks for helping out!

Please link the appropriate issue from your PR.

If you don't have an issue, we'd recommend starting with one first so the PR can focus on the
implementation (unless its an obvious bug or documentation fix that will have
little conversation).
-->

epage commented on Cargo.toml:
```diff
@@ -181,6 +181,7 @@ unstable-ext = ["clap_builder/unstable-ext"]
 unstable-styles = ["clap_builder/unstable-styles"]  # deprecated
 unstable-derive-ui-tests = []
 unstable-markdown = ["clap_derive/unstable-markdown"]
+unstable-command-headings =["unstable-command-headings"]
```
Why are we starting this off as unstable?

jerusdp replied:
I thought that was the convention.  I can rename the the feature flag or remove it if it is unnecessary.

jerusdp commented:
@epage  dropped the commits and force pushed

epage commented on clap_builder/src/output/help_template.rs:
```diff
@@ -1142,6 +1169,10 @@ fn should_show_subcommand(subcommand: &Command) -> bool {
+fn should_show_help_heading(subcommand: &Command) -> bool {
+    subcommand.is_help_heading_set()
+}
+
```
This is a pretty thin wrapper, not sure it is providing enough value

jerusdp replied:
Ok, I will remove.

epage commented on tests/builder/subcommands.rs:
```diff
@@ -635,3 +635,184 @@ fn duplicate_subcommand_alias() {
+}
+
+#[test]
+fn test_multiple_commands_mixed_standard_and_custom_headers() {
```
We don't have a test that defines commands with help headings out of order (e.g. a `First Custom`  following by a `Second Custom` followed by a `First Custom`).

jerusdp replied:
Shuffled the order of the user commands in the comprehensive mixed test.

jerusdp commented on tests/builder/subcommands.rs:
```diff
@@ -635,3 +635,184 @@ fn duplicate_subcommand_alias() {
 }
+
+#[test]
+fn test_help_header() {
```
Exposes a little of the development of thought - I have removed the first two to make the third a single comprehensive test.

epage commented on clap_builder/src/output/help_template.rs:
```diff
@@ -963,12 +967,86 @@ impl HelpTemplate<'_, '_> {
+            ord_v.insert(
+                (heading_ord, subcommand.get_display_order(), styled),
+                subcommand,
+            );
```
We should not be re-implementing `write_subcommand`

epage commented on clap_builder/src/output/help_template.rs:
```diff
@@ -931,16 +942,83 @@ impl HelpTemplate<'_, '_> {
         }
+
+        // Commands under custom headings
+        if !custom_headings.is_empty() {
```
This change is overall a lot more complicated than I would have expected and I'll likely need to dig into this a lot further to better understand how inherent the complexity is or isn't before we're able to move forward.

epage replied:
That is a fair point.  If people rely on the help headings to logically group their commands, then consistently using that order elsewhere can make sense.  There can still be room for discussing whether it is worth it or not.  Most likely, we'll need to whittle away at other complexity first to see how much of an impact it really makes.

As for the other complexity, in each case where we are dealing with custom help headings, we are duplicating all of the underlying logic rather than folding the custom and built-in heading cases together, making this more ripe for bugs.  This code also special cases `help` which will make it diverge in order compared to anything else and seems like a potential feature independent of this work.

jerusdp replied:
> > There is another concern I had last night and that is whether there are any other artefacts that are caused by the feature.
> 
> What do you mean?

That there would be some other interaction from the help_heading API - I produced a more comprehensive set of tests (all the helps tests with subcommand with a help_heading added) and only found a couple of problems with the flattened versions. One fixed and the other I am still working on.

epage commented:
> @epage dropped the commits and force pushed

Please rebase against the latest master and drop the unrelated commits

epage commented on clap_builder/src/output/help_template.rs:
```diff
@@ -892,44 +890,95 @@ impl HelpTemplate<'_, '_> {
+        self.write_flat_subcommands_under_heading(&ord_v, None, first, true);
+    }
+
+    fn write_flat_subcommands_under_heading(
```
If you are extrating logic to functions, please do that in a separate commit before this one

I'm also generally not a fan of functions inside of functions

jerusdp commented on clap_builder/src/output/help_template.rs:
```diff
@@ -389,7 +389,7 @@ impl HelpTemplate<'_, '_> {
         let subcmds = self.cmd.has_visible_subcommands();
 
-        let custom_headings = self
+        let custom_arg_headings = self
```
By a commit before you mean a commit before the change is introduced? i.e.
commit1 - above change
commit2 - implementation of help heading

gortavoher committed:
chore: Prepare for command help heading

- rename `custom_headings` to `custom_arg_headings`

gortavoher committed:
feat: Add subcommand custom headings

- add tests for subcommand headings feature
- include tests for single and multiple help headers
- add test for hiding commands header
- add test for mixed standard and custom headers
- add test case to verify that mixed headings are flattened correctly
- shuffled order in multiple command test
- add support for subcommand help headings
- display subcommands under their respective headings in help output
- only display heading if subcommand has help_heading set
- Set order for custom headings (as found)
- write commands in order under headings in order
- update usage and help templates to display in custom headings order
- introduce complex_app_with_help_heading function
- similar to complex_app but includes help_heading
- Test subcmds with help_heading
- Update tests to include the custom heading in the expected output.
- extract the logic for writing a flattened subcommand to a function

danelkis-doubleai commented:
Any updates on this one? :) We could use it
